Adding MPTCP with option to disable it and fall-back to TCP

CrapsDorian opened the pull request:
Multipath TCP (MPTCP), standardized in RFC8684 [1], is a TCP extension
that enables a TCP connection to use different paths.

Multipath TCP has been used for several use cases. On smartphones, MPTCP
enables seamless handovers between cellular and Wi-Fi networks while
preserving established connections. This use-case is what pushed Apple
to use MPTCP since 2013 in multiple applications [2]. On dual-stack
hosts, Multipath TCP enables the TCP connection to automatically use the
best performing path, either IPv4 or IPv6. If one path fails, MPTCP
automatically uses the other path.

To benefit from MPTCP, both the client and the server have to support
it. Multipath TCP is a backward-compatible TCP extension that is enabled
by default on recent Linux distributions (Debian, Ubuntu, Redhat, ...).
Multipath TCP is included in the Linux kernel since version 5.6 [3]. To
use it on Linux, an application must explicitly enable it when creating
the socket. No need to change anything else in the application.

This attached patch adds an mptcp option in the config which allows the
creation of an MPTCP socket instead of TCP on Linux. If Multipath TCP
is not supported on the system, an error will be reported and the start failed.

Due to the limited impact within a data center environment,
we have decided not to implement MPTCP between the proxy and the servers.
The high-speed, low-latency nature of data center networks reduces
the benefits of MPTCP, making the complexity of its implementation
unnecessary in this context.

Developed with the help of Matthieu Baerts (matttbe@kernel.org) and 
Olivier Bonaventure (Olivier.Bonaventure@uclouvain.be)

Link: https://www.rfc-editor.org/rfc/rfc8684.html [1]
Link: https://www.tessares.net/apples-mptcp-story-so-far/ [2]
Link: https://www.mptcp.dev/ [3]
Developed-by: Dorian Craps [doriancraps@gmail.com](mailto:doriancraps@gmail.com) @CrapsDorian
Signed-off-by: Dorian Craps [dorian.craps@student.vinci.be](mailto:dorian.craps@student.vinci.be)

matttbe commented on src/cfgparse-global.c:
```diff
@@ -51,8 +51,8 @@ static const char *common_kw_list[] = {
 	"numa-cpu-mapping", "defaults", "listen", "frontend", "backend",
-	"peers", "resolvers", "cluster-secret", "no-quic", "limited-quic",
-	NULL /* must be last */
+	"peers", "resolvers", "cluster-secret", "no-quic", "limited-quic", 
```
trailing whitespace (espace à la fin de la ligne)

matttbe commented on src/protocol.c:
```diff
@@ -104,10 +104,19 @@ int protocol_supports_flag(struct protocol *proto, uint flag)
 		/* run a live check */
-		ret = _sock_supports_reuseport(proto->fam, proto->sock_type, proto->sock_prot);
-		if (!ret)
+		if (global.tune.options & GTUNE_NO_MPTCP){
```
space needed before `{`

matttbe commented on src/protocol.c:
```diff
@@ -104,10 +104,19 @@ int protocol_supports_flag(struct protocol *proto, uint flag)
+			ret = _sock_supports_reuseport(proto->fam, proto->sock_type, proto->sock_prot);
+
+			/* retry with the fall back protocol */
+			if (!ret){
```
`{}` not needed

matttbe commented on src/protocol.c:
```diff
@@ -104,10 +104,19 @@ int protocol_supports_flag(struct protocol *proto, uint flag)
+				ret = _sock_supports_reuseport(proto->fam, proto->sock_type, proto->sock_prot_fb);
+			}
+		}
+		if (!ret){
```
`{}` not needed

matttbe commented on src/protocol.c:
```diff
@@ -104,10 +104,19 @@ int protocol_supports_flag(struct protocol *proto, uint flag)
-		ret = _sock_supports_reuseport(proto->fam, proto->sock_type, proto->sock_prot);
-		if (!ret)
+		if (global.tune.options & GTUNE_NO_MPTCP){
+			ret = _sock_supports_reuseport(proto->fam, proto->sock_type, proto->sock_prot_fb);
```
if you are not on `__linux__`, `proto->sock_prot_fb` will not be defined.

I guess you should at least have: `if (proto->sock_prot_fb > 0 && global.tune.options & GTUNE_NO_MPTCP)` (or a `#ifdef __linux__` but easier to avoid these `ifdef` I think)

matttbe commented on src/protocol.c:
```diff
@@ -104,10 +104,19 @@ int protocol_supports_flag(struct protocol *proto, uint flag)
+		}
+		if (!ret){
 			_HA_ATOMIC_AND(&proto->flags, ~PROTO_F_REUSEPORT_SUPPORTED);
-
```
keep this new line (and add one before the `if (!ret)`)

matttbe commented on src/sock_inet.c:
```diff
@@ -327,8 +327,23 @@ int sock_inet_bind_receiver(struct receiver *rx, char **errmsg)
+			* If we failed to create a socket, try to create a fallback socket
+			* with the fallback protocol.
+			*/
+			if (fd == -1) {
```
same here, only re-call `my_socketat()` if `rx->proto->sock_prot_fb > 0`

matttbe replied:
`{}` are not needed here

matttbe commented on src/sock_inet.c:
```diff
@@ -327,8 +327,23 @@ int sock_inet_bind_receiver(struct receiver *rx, char **errmsg)
 	if (!ext) {
-		fd = my_socketat(rx->settings->netns, rx->proto->fam->sock_domain,
-		                 rx->proto->sock_type, rx->proto->sock_prot);
+		if (global.tune.options & GTUNE_NO_MPTCP) {
```
same here, only if `rx->proto->sock_prot_fb > 0`

matttbe commented on src/protocol.c:
```diff
@@ -104,10 +104,19 @@ int protocol_supports_flag(struct protocol *proto, uint flag)
+
+			/* retry with the fall back protocol */
+			if (!ret){
+				ret = _sock_supports_reuseport(proto->fam, proto->sock_type, proto->sock_prot_fb);
```
Only call this if `proto->sock_prot_fb > 0`, no need to re-create the socket if there is no fallback (e.g. with UDP)

matttbe commented on src/proto_tcp.c:
```diff
@@ -80,7 +80,12 @@ struct protocol proto_tcpv4 = {
 	/* socket layer */
 	.proto_type     = PROTO_TYPE_STREAM,
 	.sock_type      = SOCK_STREAM,
+#ifdef __linux__
```
I guess you might still need to have this somewhere

```
#ifdef __linux__
#ifndef IPPROTO_MPTCP
#define IPPROTO_MPTCP
#endif
#endif
```

matttbe commented on src/protocol.c:
```diff
@@ -104,11 +104,21 @@ int protocol_supports_flag(struct protocol *proto, uint flag)
 		if (!ret)
-			_HA_ATOMIC_AND(&proto->flags, ~PROTO_F_REUSEPORT_SUPPORTED);
-
+				_HA_ATOMIC_AND(&proto->flags, ~PROTO_F_REUSEPORT_SUPPORTED);
```
wrong indentation

matttbe commented on src/protocol.c:
```diff
@@ -104,11 +104,21 @@ int protocol_supports_flag(struct protocol *proto, uint flag)
-			_HA_ATOMIC_AND(&proto->flags, ~PROTO_F_REUSEPORT_SUPPORTED);
-
+				_HA_ATOMIC_AND(&proto->flags, ~PROTO_F_REUSEPORT_SUPPORTED);
+		
```
trailing whitespaces

matttbe commented on src/sock_inet.c:
```diff
@@ -327,8 +327,23 @@ int sock_inet_bind_receiver(struct receiver *rx, char **errmsg)
-		                 rx->proto->sock_type, rx->proto->sock_prot);
+		if (rx->proto->sock_prot_fb > 0 && global.tune.options & GTUNE_NO_MPTCP) {
+			fd = my_socketat(rx->settings->netns, rx->proto->fam->sock_domain,
+							rx->proto->sock_type, rx->proto->sock_prot_fb);
```
not aligned properly: tabs up to `fd`, then spaces to align with the `(` above

matttbe commented on src/sock_inet.c:
```diff
@@ -327,8 +327,23 @@ int sock_inet_bind_receiver(struct receiver *rx, char **errmsg)
+							rx->proto->sock_type, rx->proto->sock_prot_fb);
+		} else {
+			fd = my_socketat(rx->settings->netns, rx->proto->fam->sock_domain,
+							rx->proto->sock_type, rx->proto->sock_prot);
```
same here

matttbe commented on src/sock_inet.c:
```diff
@@ -327,8 +327,23 @@ int sock_inet_bind_receiver(struct receiver *rx, char **errmsg)
+			*/
+			if (rx->proto->sock_prot_fb > 0 && fd == -1) 
+				fd = my_socketat(rx->settings->netns, rx->proto->fam->sock_domain,
+							rx->proto->sock_type, rx->proto->sock_prot_fb);
```
same here

matttbe commented on src/protocol.c:
```diff
@@ -104,11 +104,21 @@ int protocol_supports_flag(struct protocol *proto, uint flag)
+			/* retry with the fall back protocol */
+			if (proto->sock_prot_fb > 0 && !ret)
+				ret = _sock_supports_reuseport(proto->fam, proto->sock_type, proto->sock_prot_fb);
+			
```
remove blank link

matttbe commented on src/protocol.c:
```diff
@@ -104,11 +104,21 @@ int protocol_supports_flag(struct protocol *proto, uint flag)
+
+			/* retry with the fall back protocol */
+			if (proto->sock_prot_fb > 0 && !ret)
+				ret = _sock_supports_reuseport(proto->fam, proto->sock_type, proto->sock_prot_fb);
```
you might need to go to the new line after ~80 chars (or 100?)

matttbe commented on src/sock_inet.c:
```diff
@@ -327,8 +327,23 @@ int sock_inet_bind_receiver(struct receiver *rx, char **errmsg)
+			if (rx->proto->sock_prot_fb > 0 && fd == -1) 
+				fd = my_socketat(rx->settings->netns, rx->proto->fam->sock_domain,
+							rx->proto->sock_type, rx->proto->sock_prot_fb);
+			
```
remove blank line here before `}`

matttbe reviewed (commented):
Please check https://github.com/CrapsDorian/haproxy/blob/master/doc/coding-style.txt

CrapsDorian committed:
FEATURE: Adding MPTCP with option to disable it and fall-back to TCP

Multipath TCP (MPTCP), standardized in RFC8684 [1], is a TCP extension
that enables a TCP connection to use different paths.

Multipath TCP has been used for several use cases. On smartphones, MPTCP
enables seamless handovers between cellular and Wi-Fi networks while
preserving established connections. This use-case is what pushed Apple
to use MPTCP since 2013 in multiple applications [2]. On dual-stack
hosts, Multipath TCP enables the TCP connection to automatically use the
best performing path, either IPv4 or IPv6. If one path fails, MPTCP
automatically uses the other path.

To benefit from MPTCP, both the client and the server have to support
it. Multipath TCP is a backward-compatible TCP extension that is enabled
by default on recent Linux distributions (Debian, Ubuntu, Redhat, ...).
Multipath TCP is included in the Linux kernel since version 5.6 [3]. To
use it on Linux, an application must explicitly enable it when creating
the socket. No need to change anything else in the application.

This attached patch adds an mptcp option in the config which allows the
creation of an MPTCP socket instead of TCP on Linux. If Multipath TCP
is not supported on the system, an error will be reported and the start failed.

Due to the limited impact within a data center environment,
we have decided not to implement MPTCP between the proxy and the servers.
The high-speed, low-latency nature of data center networks reduces
the benefits of MPTCP, making the complexity of its implementation
unnecessary in this context.

Developed with the help of Matthieu Baerts (matttbe@kernel.org) and 
Olivier Bonaventure (Olivier.Bonaventure@uclouvain.be)

Link: https://www.rfc-editor.org/rfc/rfc8684.html [1]
Link: https://www.tessares.net/apples-mptcp-story-so-far/ [2]
Link: https://www.mptcp.dev/ [3]
Developed-by: Dorian Craps [doriancraps@gmail.com](mailto:doriancraps@gmail.com) @CrapsDorian
Signed-off-by: Dorian Craps [dorian.craps@student.vinci.be](mailto:dorian.craps@student.vinci.be)

matttbe commented on src/protocol.c:
```diff
@@ -104,11 +104,22 @@ int protocol_supports_flag(struct protocol *proto, uint flag)
+
 		if (!ret)
-			_HA_ATOMIC_AND(&proto->flags, ~PROTO_F_REUSEPORT_SUPPORTED);
+			HA_ATOMIC_AND(&proto->flags, ~PROTO_F_REUSEPORT_SUPPORTED);
```
it looks like you accidentally removed a `_` at the beginning

matttbe commented on src/protocol.c:
```diff
@@ -104,11 +104,22 @@ int protocol_supports_flag(struct protocol *proto, uint flag)
+			HA_ATOMIC_AND(&proto->flags, ~PROTO_F_REUSEPORT_SUPPORTED);
 
 		_HA_ATOMIC_OR(&proto->flags, PROTO_F_REUSEPORT_TESTED);
+
```
best not to modify this part of the code you didn't change: remove the blank line here

matttbe commented on doc/configuration.txt:
```diff
@@ -1341,6 +1341,7 @@ The following keywords are supported in the "global" section :
    - maxsslconn
    - maxsslrate
    - maxzlibmem
```
commit msg:

> This attached patch adds an mptcp option in the config which allows the
> creation of an MPTCP socket instead of TCP on Linux. If Multipath TCP
> is not supported on the system, an error will be reported and the start failed.

That's not correct here: MPTCP is used by default (you need to explain why it is OK to do that here for the server case: you can quote the [website](https://www.mptcp.dev/faq.html#why--when-should-mptcp-be-enabled-by-default) and add a ref to it) if it is supported by the system. If not, TCP is used like before.

You also need to mention:
- `IPPROTO_MPTCP` is defined just in case old libC are being used and don't have the ref. The running kernel is the only one who can tell if MPTCP is supported or not.
- `TCP_MAXSEG` is currently not supported by MPTCP: is it an issue? MPTCP devs didn't add a support for it because it has not been requested with a use-case, but it can be added.

matttbe commented on doc/configuration.txt:
```diff
@@ -2958,6 +2959,9 @@ maxzlibmem <number>
   with "show info" on the line "MaxZlibMemUsage", the memory used by zlib is
   "ZlibMemUsage" in bytes.
 
+no-mptcp
```
same here: should be moved under `no-memory-trimming` if it is by alphabetical order

matttbe commented on doc/configuration.txt:
```diff
@@ -1341,6 +1341,7 @@ The following keywords are supported in the "global" section :
    - maxsslconn
    - maxsslrate
    - maxzlibmem
+   - no-mptcp
```
should be moved under `no-memory-trimming` if it is by alphabetical order

matttbe commented on doc/configuration.txt:
```diff
@@ -2958,6 +2959,9 @@ maxzlibmem <number>
   "ZlibMemUsage" in bytes.
 
+no-mptcp
+  Disables the use of the Multipath TCP (MPTCP) protocol to use TCP protocol.
```
Maybe:

```
Disables Multipath TCP (MPTCP) support when the TCP protocol is requested.
```

matttbe commented on src/sock_inet.c:
```diff
@@ -327,8 +327,22 @@ int sock_inet_bind_receiver(struct receiver *rx, char **errmsg)
+			* If we failed to create a socket, try to create a fallback socket
+			* with the fallback protocol.
+			*/
+			if (rx->proto->sock_prot_fb > 0 && fd == -1) 
```
I wonder if it might not be interesting to log something here:

```
memprintf(errmsg, "protocol %d not supported, falling back to %d",
          rx->proto->sock_prot, rx->proto->sock_prot_fb);
err |= ERR_ALERT;
```

matttbe reviewed (changes requested):
almost there!
